File: demos/typescript/geometry-samples/src/samples/terrain.ts
// Terrain as the graph of a scalar field.
//
// Two stdlib pieces meet here. `ValueNoise2D` is the stdlib's value noise:
// one hashed value per integer lattice corner, faded between them. Summing
// octaves at doubling frequency and halving amplitude is fractal Brownian
// motion; the sum is wrapped in `ScalarFunctionField2D`, the stdlib's
// function-backed scalar field, which makes it an IScalarField2D like any
// other. `ToTriangleMesh(domain, nCols, nRows)` then draws any such field as
// the surface z = f(x, y).
//
// fBm is composed here rather than using the stdlib's own `FbmNoise2D` because
// that type's Basis field is the sum type NoiseBasis, which has no TypeScript
// surface (CHK320) — see plato-440.

import type { Drawable, Sample } from '../core/types.js';
import {
    Bounds2D, Point2D, ScalarFunctionField2D, ValueNoise2D,
} from '../plato/plato.g.js';

const octave = ValueNoise2D.Create(1234, 1);

/** Stdlib value noise at (x, y), in [-1, 1]. */
export const valueNoise = (x: number, y: number): number =>
    octave.Eval(new Point2D(x, y));

/** Fractal Brownian motion: octaves at doubling frequency, halving amplitude. */
export function fbm(p: Point2D, octaves = 5): number {
    let sum = 0, amplitude = 0.5, frequency = 1, total = 0;
    for (let i = 0; i < octaves; i++) {
        sum += amplitude * valueNoise(p.X * frequency, p.Y * frequency);
        total += amplitude;
        amplitude *= 0.5;
        frequency *= 2;
    }
    return sum / total; // back to [-1, 1]
}

/** The fBm sum as a stdlib scalar field, scaled to the terrain's relief. */
export const terrainField = ScalarFunctionField2D.Create(p => fbm(p) * 0.45);

export const terrainSample: Sample = {
    id: 'terrain',
    title: 'Value-Noise Terrain',
    description: 'ScalarFunctionField2D over stdlib ValueNoise2D, drawn as its graph ' +
        'by IScalarField2D.ToTriangleMesh.',
    build(): Drawable[] {
        const domain = new Bounds2D(new Point2D(-1.6, -1.6), new Point2D(1.6, 1.6));
        return [{
            kind: 'mesh',
            mesh: terrainField.ToTriangleMesh(domain, 140, 140),
            color: 0x6fbf73,
            flatShading: true,
        }];
    },
};

File: demos/rust/geometry-samples/src/samples/terrain.rs
// Heightfield terrain from fractal value noise.
//
// Value noise: hash the integer lattice, then interpolate with a smoothstep.
// Fractal Brownian motion (fBm) sums several octaves at doubling frequency
// and halving amplitude. The result displaces a grid mesh vertically.

use crate::core::mesh_builder::grid_mesh;
use crate::core::types::Drawable;
use crate::plato::*;

/// Deterministic pseudo-random value in [0, 1) for an integer lattice point.
fn lattice_hash(ix: f64, iz: f64) -> f64 {
    let h = (ix * 127.1 + iz * 311.7).sin() * 43758.5453;
    h - h.Floor()
}

fn smoothstep(t: f64) -> f64 {
    t * t * (3.0 - 2.0 * t)
}

pub fn value_noise(x: f64, z: f64) -> f64 {
    let (ix, iz) = (x.Floor(), z.Floor());
    let (fx, fz) = (smoothstep(x - ix), smoothstep(z - iz));
    let v00 = lattice_hash(ix, iz);
    let v10 = lattice_hash(ix + 1.0, iz);
    let v01 = lattice_hash(ix, iz + 1.0);
    let v11 = lattice_hash(ix + 1.0, iz + 1.0);
    v00.Lerp(v10, fx).Lerp(v01.Lerp(v11, fx), fz) // bilinear, in [0, 1)
}

pub fn fbm(x: f64, z: f64, octaves: usize) -> f64 {
    let (mut sum, mut amplitude, mut frequency, mut total) = (0.0, 0.5, 1.0, 0.0);
    for _ in 0..octaves {
        sum += amplitude * value_noise(x * frequency, z * frequency);
        total += amplitude;
        amplitude *= 0.5;
        frequency *= 2.0;
    }
    sum / total // normalized back to [0, 1)
}

pub fn build() -> Vec<Drawable> {
    let (size, height) = (3.2, 0.9);
    let mesh = grid_mesh(120, 120, |u, v| {
        Vector3D::new(
            (u - 0.5) * size,
            fbm(u * 5.0, v * 5.0, 4) * height - height * 0.5,
            (v - 0.5) * size,
        )
    });
    vec![mesh.into_drawable()]
}
